File: src/main/java/com/blackcowmoo/moomark/auth/util/AesUtil.java
package com.blackcowmoo.moomark.auth.util;

import java.security.InvalidKeyException;
import javax.crypto.BadPaddingException;
import javax.crypto.IllegalBlockSizeException;
import javax.crypto.NoSuchPaddingException;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Component;

import javax.crypto.Cipher;
import javax.crypto.KeyGenerator;
import javax.crypto.SecretKey;
import java.security.NoSuchAlgorithmException;
import java.security.SecureRandom;

@Component
@Slf4j
public class AesUtil {

  private AesUtil() throws InstantiationException {
    throw new InstantiationException("Can not instantiation class");
  }

  private static final String CIPHER = "AES";
  private static final Integer KEY_SIZE = 128;
  private static final SecureRandom secureRandom = new SecureRandom();

  public static byte[] encrypt(String body, SecretKey key)
    throws NoSuchAlgorithmException, InvalidKeyException,
    IllegalBlockSizeException, BadPaddingException, NoSuchPaddingException {
    Cipher cipher = Cipher.getInstance(CIPHER);
    cipher.init(Cipher.ENCRYPT_MODE, key);
    return cipher.doFinal(body.getBytes());
  }

  public static String decrypt(byte[] body, SecretKey key)
    throws NoSuchAlgorithmException, InvalidKeyException,
    IllegalBlockSizeException, BadPaddingException, NoSuchPaddingException {
    Cipher cipher = Cipher.getInstance(CIPHER);
    cipher.init(Cipher.DECRYPT_MODE, key);
    byte[] decrypted = cipher.doFinal(body);
    return new String(decrypted);
  }

  public static SecretKey generateNewKey() {
    return getRandomKey();
  }

  private static SecretKey getRandomKey() {
    KeyGenerator generator = null;
    try {
      generator = KeyGenerator.getInstance(AesUtil.CIPHER);
      SecureRandom random = SecureRandom.getInstance("SHA1PRNG");
      random.setSeed(getRandomBytes());
      generator.init(KEY_SIZE, random);
      return generator.generateKey();
    } catch (NoSuchAlgorithmException e) {
      throw new IllegalStateException(e.getMessage(), e);
    }
  }

  private static byte[] getRandomBytes() {
    byte[] secureRandomKeyBytes = new byte[AesUtil.KEY_SIZE / 8];
    secureRandom.nextBytes(secureRandomKeyBytes);
    return secureRandomKeyBytes;
  }
}

File: src/main/kotlin/com/blackcowmoo/moomark/auth/util/AesUtil.kt
package com.blackcowmoo.moomark.auth.util

import lombok.extern.slf4j.Slf4j
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component
import java.security.NoSuchAlgorithmException
import java.security.SecureRandom
import javax.crypto.Cipher
import javax.crypto.KeyGenerator
import javax.crypto.SecretKey

@Component
@Slf4j
class AesUtil {
  companion object {
    private val log = LoggerFactory.getLogger(AesUtil::class.java)
  }
  private val cipher = "AES"
  private val secureRandom = SecureRandom()

  @Throws(Exception::class)
  fun encrypt(body: String, key: SecretKey): ByteArray {
    val cipher = Cipher.getInstance("AES")
    cipher.init(Cipher.ENCRYPT_MODE, key)
    val encrypted = cipher.doFinal(body.toByteArray())
    return encrypted
  }

  fun decrypt(body: ByteArray, key: SecretKey): String {
    val cipher = Cipher.getInstance("AES")
    cipher.init(Cipher.DECRYPT_MODE, key)
    val decrypted = cipher.doFinal(body)
    return String(decrypted, Charsets.UTF_8)
  }

  fun generateNewKey(): SecretKey? {
    return getRandomKey(cipher, 128)
  }

  private fun getRandomKey(cipher: String, keySize: Int): SecretKey? {
    return try {
      val generator = KeyGenerator.getInstance(cipher)
      val random = SecureRandom.getInstance("SHA1PRNG")
      random.setSeed(getRandomBytes(keySize))
      generator.init(128, random)
      generator.generateKey()
    } catch (e: NoSuchAlgorithmException) {
      log.error(e.message, e)
      null
    }
  }

  private fun getRandomBytes(keySize: Int): ByteArray {
    val secureRandomKeyBytes = ByteArray(keySize / 8)
    secureRandom.nextBytes(secureRandomKeyBytes)
    return secureRandomKeyBytes
  }
}
